### PR TITLE
Remove duplicate class analysis in incremental compiler

### PR DESCRIPTION
**Pull request by oehme**
This reworks the task-specific caches of the incremental compiler to be simpler and contain less data. The class analysis for a task's output directory is no longer kept in memory. No one was reusing it, so there was no reason to keep it around. Instead, it is now computed on demand when the incremental compiler needs to figure out the affected classes. There is now only one code path for class analysis instead of two, greatly simplifying the code.

### CODE REVIEW

**Comment by wolfs** [comment]
@oehme We invalidate the file system mirror before the task starts since from that point in time the task can change something on disk and the entries in the mirror may not be valid any more.

If you want to obtain the snapshot for a classpath entry, why don't you use the `FileCollectionSnapshot` for the classpath property which has been already calculated? Does the snapshot not have sufficiently fine granular information for you? The `FileCollectionSnapshot` is part of `CurrentExecution` in the `TaskHistoryRepository.History`.

**Comment by oehme** [comment]
@wolfs The problem is not snapshotting a classpath entry. The problem is getting a snapshot of my previous output (which others may have used as a classpath entry). But your suggestion got me in the right direction I think - I'll try `HistoricalTaskExecution` 👍

**Comment by oehme** [comment]
It's actually pretty hard to get a hold of the `History` object from a task. So for now I left out the cache reuse and just removed the duplicate memory usage. This means we're still spending some unnecessary CPU cycles until I figure out a way to get to that history. @wolfs any pointers for that?

**Comment by oehme** [comment]
I just did some heapdumps locally and the improvement was well worth the effort. On Gradle 4.9, the incremental compiler added 250MB of heap to the `largeJavaMultiProject`. On this branch the overhead is down to 30MB.

File: subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/IncrementalCompilerFactory.java [hunk 1]
```diff
@@ -49,41 +50,39 @@
...
-    public IncrementalCompilerFactory(FileOperations fileOperations, StreamHasher streamHasher, FileHasher fileHasher, GeneralCompileCaches generalCompileCaches, BuildOperationExecutor buildOperationExecutor, StringInterner interner, FileSystemSnapshotter fileSystemSnapshotter) {
+    public IncrementalCompilerFactory(FileOperations fileOperations, StreamHasher streamHasher, GeneralCompileCaches generalCompileCaches, BuildOperationExecutor buildOperationExecutor, StringInterner interner, FileSystemSnapshotter fileSystemSnapshotter, FileHasher fileHasher) {
         this.fileOperations = fileOperations;
         this.streamHasher = streamHasher;
-        this.fileHasher = fileHasher;
         this.generalCompileCaches = generalCompileCaches;
         this.buildOperationExecutor = buildOperationExecutor;
         this.interner = interner;
         this.fileSystemSnapshotter = fileSystemSnapshotter;
+        this.fileHasher = fileHasher;
```
Comment (marcphilipp) [inline subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/IncrementalCompilerFactory.java]: Nit: there's no need to change the argument order of the constructor, is there?

Comment (oehme) [reply subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/IncrementalCompilerFactory.java]: No, this was removed and later added again :) But it doesn't hurt either, so I'm gonna spare our CI another round.

**Pull request merged**